File: python/hermetix/brokers/next.py
"""넥스트증권 모의투자 어댑터 (미국주식) — 공개 스펙 v1.3 기준.

v1.3 응답(quotes outcome, 캔들 time, 계좌 cashAmount, 보유 averageBuyPrice, 캘린더 status+sessions[] …)을
브로커 중립 공통 모델(Quote/Candle/Holding/Order …)로 정규화한다. 공통 모델은 전략이 보는 타입이므로
서버 스펙 변경은 이 어댑터 안에서만 흡수한다 (KIS/키움 어댑터와 같은 방식).

- OAuth client_credentials, 토큰 12h(expires_in=43200). 401 시 1회 재발급-재시도
- 공통 헤더(v1.3): X-Request-Id 는 토큰 발급 외 전 API 필수(누락 시 400 request-id-required),
  계좌·자산·주문 API 는 X-Next-Account-Id 필수(구 X-Nextsecurities-Account 에서 개명)
- 시각은 ISO 8601 · KST (오프셋 생략 시 KST). 등락률·손익률은 % 단위 → 공통 모델 규약(비율)로 /100
- 에러 엔벨로프: {"error": {type, code, message, ...}}. 토큰 발급 400/401 만 OAuth 표준 {error, error_description}
- 고급 주문(STOP/BRACKET)/kill-switch/modify/cancel-all 은 /v2 로 제공 — 이 어댑터는 아직 미연동
"""
from __future__ import annotations

import re
import threading
import time
import uuid
from datetime import datetime, timezone
from decimal import Decimal
from zoneinfo import ZoneInfo

from ..broker import BrokerClient, _Http
from ..errors import (
    AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError,
    MarketClosedError, OrderNotFoundError, RateLimitError,
)
from ..models import (
    Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
    Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, OrderType, Quote,
    SessionHours, TradingEnvironment,
)

ACCOUNT_HEADER = "X-Next-Account-Id"
REQUEST_ID_HEADER = "X-Request-Id"
_REQUEST_ID_PATTERN = re.compile(r"^[A-Za-z0-9._-]{1,64}$")

KST = ZoneInfo("Asia/Seoul")
NEW_YORK = ZoneInfo("America/New_York")
_MARKET = "US"


def new_request_id() -> str:
    """v1.3 규칙: 영숫자·점·밑줄·하이픈만, 최대 64자. hmx- 프리픽스 + UUID = 40자."""
    rid = f"hmx-{uuid.uuid4()}"
    assert _REQUEST_ID_PATTERN.match(rid)
    return rid


def _d(value) -> Decimal | None:
    if value is None or value == "":
        return None
    return Decimal(str(value))


def _pct(value) -> Decimal | None:
    """% 단위 → 비율 (3.3333 → 0.033333)"""
    rate = _d(value)
    return None if rate is None else rate / Decimal(100)


def _ts(value: str | None) -> datetime | None:
    """v1.3 시각: ISO 8601, 오프셋 생략 시 KST"""
    if not value:
        return None
    parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=KST)
    return parsed


def _ny_clock(value: str | None) -> str | None:
    """KST ISO 시각 → 뉴욕 현지 HH:MM (공통 캘린더 모델은 현지 타임존 + HH:MM)"""
    parsed = _ts(value)
    return None if parsed is None else parsed.astimezone(NEW_YORK).strftime("%H:%M")


class NextClient(BrokerClient):

    capabilities = BrokerCapabilities(
        broker_id="next",
        market="US",
        currency="USD",
        candle_intervals=frozenset({CandleInterval.MIN_1, CandleInterval.DAY_1}),  # v1.3: 1m · 1d
        client_order_id=True,
        native_bracket=False,  # 서버 /v2/orders/advanced(BRACKET) 연동 전까지 소프트웨어 브라켓
        fractional_shares=False,  # v1.3 주문 수량은 정수만
        server_open_orders=True,
        environments=frozenset({TradingEnvironment.PAPER, TradingEnvironment.LIVE}),  # 키 프리픽스로 결정
    )

    def __init__(self, client_id: str, client_secret: str, account_id: str = "acc_main",
                 base_url: str = "https://openapi.nextsecurities.dev",
                 environment: TradingEnvironment = TradingEnvironment.PAPER):
        # 환경은 키 프리픽스가 결정한다 (pk_test_=모의, pk_live_=실전) — 설정과 어긋나면 기동 실패
        expected = "pk_live_" if environment == TradingEnvironment.LIVE else "pk_test_"
        if client_id.startswith("pk_") and not client_id.startswith(expected):
            raise ValueError(f"environment={environment.value} 인데 client_id 가 '{expected}' 로 시작하지 않습니다 "
                             "(모의=pk_test_, 실전=pk_live_). 키와 환경 설정을 맞추세요.")
        self.environment = environment
        self._client_id = client_id
        self._client_secret = client_secret
        self._account_id = account_id
        self._http = _Http(base_url)
        self._token: str | None = None
        self._token_expires_at = 0.0
        self._token_lock = threading.Lock()

    # ------------------------------------------------------------------ market

    def get_quotes(self, symbols: list[str]) -> list[Quote]:
        requested = {self.capabilities.symbol_code(s): s for s in symbols}
        body = self._get("/v1/market/quotes", query={"symbols": ",".join(requested)})
        quotes = []
        for q in body.get("quotes", []):
            # NOT_FOUND / NO_DATA 는 개별 종목의 정상 결과 — 가격이 없으므로 제외한다 (ctx.quote() 가 None)
            if q.get("outcome") != "OK" or q.get("price") is None:
                continue
            quotes.append(Quote(
                symbol=requested.get(q["symbol"], q["symbol"]),  # 요청받은 표기(시장 접두 포함)로
                price=_d(q["price"]),
                bid_price=_d(q.get("bidPrice")),
                ask_price=_d(q.get("askPrice")),
                volume=int(q.get("volume") or 0),
                change=_d(q.get("change")),
                change_rate=_pct(q.get("changeRate")),
                timestamp=_ts(q.get("lastTradeAt")) or _ts(q.get("requestedAt")) or datetime.now(timezone.utc),
            ))
        return quotes

    def get_candles(self, symbol: str, interval: CandleInterval, limit: int | None = None) -> list[Candle]:
        query = {"symbol": self.capabilities.symbol_code(symbol), "interval": interval.value}
        if limit is not None:
            query["limit"] = str(limit)
        body = self._get("/v1/market/candles", query=query)
        return [
            Candle(
                timestamp=_ts(c["time"]),
                open=_d(c["open"]), high=_d(c["high"]), low=_d(c["low"]), close=_d(c["close"]),
                volume=int(c.get("volume") or 0),
            )
            for c in body.get("candles", [])
        ]

    def get_calendar(self) -> list[MarketDay]:
        """v1.3: date 는 거래소 현지 일자, 세션 시각은 KST → 뉴욕 현지 HH:MM 으로 바꿔 담는다."""
        body = self._get("/v1/market/calendar")
        days = []
        for d in body.get("calendar", []):
            regular = next((s for s in d.get("sessions") or [] if s.get("type") == "REGULAR"), None)
            start = _ny_clock(regular.get("open")) if regular else None
            end = _ny_clock(regular.get("close")) if regular else None
            is_open = d.get("status") in ("OPEN", "HALF_DAY") and start is not None and end is not None
            days.append(MarketDay(
                date=d["date"],
                open=is_open,
                regular=SessionHours(start, end) if is_open else None,
                timezone="America/New_York",
                holiday=d.get("holidayName"),
            ))
        return days

    # ----------------------------------------------------------------- account

    def get_account(self) -> Account:
        """v1.3 계좌 응답은 예수금(cashAmount)만 준다. 총평가는 예수금 + 보유 평가금액 합 (보유 조회 1회 추가)."""
        body = self._get("/v1/account", account=True)
        cash = _d(body["cashAmount"])
        market_value = sum((h.market_value or Decimal(0) for h in self.get_holdings()), Decimal(0))
        return Account(
            account_id=body["accountId"],
            currency=body.get("currency") or "USD",
            cash=cash,
            portfolio_value=cash + market_value,
            status="ACTIVE",
            name=None,
        )

    def get_holdings(self) -> list[Holding]:
        body = self._get("/v1/account/holdings", account=True)
        return [
            Holding(
                symbol=h["symbol"],
                quantity=_d(h["quantity"]),
                avg_entry_price=_d(h["averageBuyPrice"]),
                current_price=_d(h.get("currentPrice")),
                market_value=_d(h.get("evaluationAmount")),
                unrealized_pnl=_d(h.get("evaluationPnl")),
                unrealized_pnl_rate=_pct(h.get("evaluationPnlRate")),
            )
            for h in body.get("holdings", [])
        ]

    def get_buying_power(self) -> Decimal:
        return _d(self._get("/v1/account/buying-power", account=True)["buyingPower"])

    # ------------------------------------------------------------------ orders

    def create_order(self, request: CreateOrderRequest) -> Order:
        payload = {
            # v1.3: clientOrderId(멱등키)·market 필수 — 호출자가 안 주면 어댑터가 UUID 를 만든다
            "clientOrderId": request.client_order_id or str(uuid.uuid4()),
            "market": _MARKET,
            "symbol": self.capabilities.symbol_code(request.symbol),
            "side": request.side.value,
            "orderType": request.order_type.value,
            "quantity": str(request.quantity),
            "timeInForce": request.time_in_force.value,
        }
        if request.limit_price is not None:
            payload["limitPrice"] = str(request.limit_price)
        return self._order(self._post("/v1/orders", payload, account=True))

    def get_orders(self) -> list[Order]:
        body = self._get("/v1/orders", account=True)
        return [self._order(o) for o in body.get("orders", [])]

    def get_order(self, order_id: str) -> Order:
        return self._order(self._get(f"/v1/orders/{order_id}", account=True))

    def cancel_order(self, order_id: str) -> Order:
        return self._order(self._request("DELETE", f"/v1/orders/{order_id}", account=True))

    def get_fills(self) -> list[Fill]:
        body = self._get("/v1/orders/fills", account=True)
        return [
            Fill(
                fill_id=None,  # v1.3: 원장이 체결 ID 를 발급하지 않는다
                order_id=f.get("orderId"), symbol=f.get("symbol"),
                side=OrderSide(f["side"]) if f.get("side") in ("BUY", "SELL") else None,
                quantity=_d(f.get("quantity")), price=_d(f.get("price")),
            )
            for f in body.get("fills", [])
        ]

    # ---------------------------------------------------------------- internal

    @staticmethod
    def _order(body: dict) -> Order:
        """생성/상세/취소 응답 공통 — 생성·취소는 orderId/status/requestedAt 만 온다."""
        status = body.get("status", "UNKNOWN")
        order_type = (body.get("orderType") or "").upper()
        return Order(
            order_id=body["orderId"],
            status=OrderStatus(status) if status in OrderStatus.__members__ else OrderStatus.UNKNOWN,
            symbol=body.get("symbol"),
            side=OrderSide(body["side"]) if body.get("side") in ("BUY", "SELL") else None,
            order_type=OrderType(order_type) if order_type in ("MARKET", "LIMIT") else None,
            quantity=_d(body.get("quantity")),
            limit_price=_d(body.get("limitPrice")),
            filled_quantity=_d(body.get("filledQuantity")),
            avg_fill_price=_d(body.get("avgFillPrice")),
            client_order_id=body.get("requestId"),  # v1.3 상세 조회의 clientOrderId 필드명
            submitted_at=_ts(body.get("requestedAt")),
            canceled_at=None,
        )

    def _get(self, path: str, *, query: dict | None = None, account: bool = False) -> dict:
        return self._request("GET", path, query=query, account=account)

    def _post(self, path: str, payload: dict, *, account: bool = False) -> dict:
        return self._request("POST", path, json_body=payload, account=account)

    def _request(self, method: str, path: str, *, query: dict | None = None,
                 json_body: dict | None = None, account: bool = False) -> dict:
        def call() -> dict:
            headers = {
                "Authorization": f"Bearer {self._get_token()}",
                REQUEST_ID_HEADER: new_request_id(),
            }
            if account and self._account_id:
                headers[ACCOUNT_HEADER] = self._account_id
            status, body = self._http.request(method, path, headers=headers, query=query, json_body=json_body)
            if status < 200 or status >= 300:
                raise self._map_error(status, body.get("error") or {})
            return body

        try:
            return call()
        except AuthError:
            # 토큰 만료 - 1회 재발급 후 재시도
            self._token = None
            return call()

    @staticmethod
    def _map_error(status: int, error: dict) -> BrokerApiError:
        """v1.3 에러 type ↔ HTTP: validation(400) authentication(401) permission(403) not_found(404)
        conflict(409) business_rule(422) locked(423, 킬스위치 trading-halted) rate_limit(429) server(5xx)"""
        code = error.get("code")
        error_type = error.get("type")
        message = f"Next({code}) {error.get('message', '')} requestId={error.get('requestId')}"
        if status == 401 or error_type == "authentication":
            return AuthError(status, code, message)
        if status == 429:
            return RateLimitError(status, code, message)
        if error_type == "permission":
            # 조회전용 키(insufficient-scope)·계좌 불일치 — 자금 부족으로 오인하지 않는다
            return BrokerApiError(status, code, message)
        if code == "order-not-found":
            return OrderNotFoundError(code, message)
        if code and "insufficient" in code:
            return InsufficientFundsError(status, code, message)
        if code == "trading-halted" or (code and "market-closed" in code):
            return MarketClosedError(status, code, message)
        if error_type in ("validation", "business_rule"):
            return InvalidOrderError(status, code, message)
        return BrokerApiError(status, code, message)

    def _get_token(self) -> str:
        if self._token and time.time() < self._token_expires_at - 60:
            return self._token
        with self._token_lock:
            if self._token and time.time() < self._token_expires_at - 60:
                return self._token
            status, body = self._http.request(
                "POST", "/v1/oauth/token",
                headers={REQUEST_ID_HEADER: new_request_id()},
                form_body={"grant_type": "client_credentials",
                           "client_id": self._client_id, "client_secret": self._client_secret},
            )
            if status != 200 or "access_token" not in body:
                raise self._token_error(status, body)
            self._token = body["access_token"]
            self._token_expires_at = time.time() + float(body.get("expires_in", 43200))
            return self._token

    @staticmethod
    def _token_error(status: int, body: dict) -> AuthError:
        error = body.get("error")
        # 토큰 발급 400/401 은 OAuth 표준 형식: {"error": "invalid_client", "error_description": "..."}
        if isinstance(error, str):
            return AuthError(status, error, f"Next 토큰 발급 실패({error}): {body.get('error_description') or ''}")
        # 429/5xx 는 플랫폼 엔벨로프: {"error": {"code": .., "message": ..}}
        error = error or {}
        return AuthError(status, error.get("code"), f"Next 토큰 발급 실패({error.get('code')}): {error.get('message') or ''}")

File: js/src/brokers/next.ts
/**
 * 넥스트증권 모의투자 어댑터 (미국주식) — 공개 스펙 v1.3 기준.
 *
 * v1.3 응답(quotes outcome, 캔들 time, 계좌 cashAmount, 보유 averageBuyPrice, 캘린더 status+sessions[] …)을
 * 브로커 중립 공통 모델(Quote/Candle/Holding/Order …)로 정규화한다. 공통 모델은 전략이 보는 타입이므로
 * 서버 스펙 변경은 이 어댑터 안에서만 흡수한다 (KIS/키움 어댑터와 같은 방식).
 *
 * - OAuth client_credentials, 토큰 12h(expires_in=43200), 401 시 1회 재발급-재시도
 * - 공통 헤더(v1.3): X-Request-Id 는 토큰 발급 외 전 API 필수, 계좌·자산·주문 API 는 X-Next-Account-Id 필수
 *   (구 X-Nextsecurities-Account 에서 개명)
 * - 시각은 ISO 8601 · KST (오프셋 생략 시 KST). 등락률·손익률은 % 단위 → 공통 모델 규약(비율)로 /100
 * - 토큰 발급 400/401 만 OAuth 표준 {error, error_description} 형식
 */
import { randomUUID } from "node:crypto";
import { Decimal } from "decimal.js";
import { BrokerClient, D, DorNull, httpJson } from "../broker.js";
import {
  AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError,
  MarketClosedError, OrderNotFoundError, RateLimitError,
} from "../errors.js";
import type {
  Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
  Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, OrderType, Quote,
} from "../models.js";

interface NextError { type?: string; code?: string; message?: string; requestId?: string; }

export const NEXT_ACCOUNT_HEADER = "X-Next-Account-Id";
export const NEXT_REQUEST_ID_HEADER = "X-Request-Id";

/** v1.3 규칙: 영숫자·점·밑줄·하이픈만, 최대 64자. hmx- 프리픽스 + UUID = 40자 */
export const newRequestId = (): string => `hmx-${randomUUID()}`;

const MARKET = "US";
const NEW_YORK = "America/New_York";
const ORDER_STATUSES = ["SUBMITTED", "PARTIALLY_FILLED", "PENDING_CANCEL", "FILLED", "CANCELED", "REJECTED", "EXPIRED"];

/** v1.3 시각: ISO 8601, 오프셋 생략 시 KST */
export function parseKst(value: unknown): Date | null {
  if (typeof value !== "string" || value.trim() === "") return null;
  const hasOffset = /(Z|[+-]\d{2}:?\d{2})$/i.test(value);
  const date = new Date(hasOffset ? value : `${value}+09:00`);
  return Number.isNaN(date.getTime()) ? null : date;
}

/** KST ISO 시각 → 뉴욕 현지 HH:mm (공통 캘린더 모델은 현지 타임존 + HH:mm) */
function nyClock(value: unknown): string | null {
  const date = parseKst(value);
  if (!date) return null;
  return new Intl.DateTimeFormat("en-GB", {
    timeZone: NEW_YORK, hour: "2-digit", minute: "2-digit", hourCycle: "h23",
  }).format(date);
}

/** % 단위 → 비율 (3.3333 → 0.033333) */
const pct = (value: unknown): Decimal | null => {
  const rate = DorNull(value);
  return rate ? rate.div(100) : rate;
};

export class NextClient implements BrokerClient {
  readonly capabilities: BrokerCapabilities = {
    brokerId: "next",
    market: "US",
    currency: "USD",
    candleIntervals: new Set<CandleInterval>(["1m", "1d"]), // v1.3: 1m · 1d
    clientOrderId: true,
    nativeBracket: false, // 서버 /v2/orders/advanced(BRACKET) 연동 전까지 소프트웨어 브라켓
    fractionalShares: false, // v1.3 주문 수량은 정수만
    serverOpenOrders: true,
  };

  private token: string | null = null;
  private tokenExpiresAt = 0;

  constructor(
    private readonly clientId: string,
    private readonly clientSecret: string,
    private readonly accountId: string = "acc_main",
    private readonly baseUrl: string = "https://openapi.nextsecurities.dev",
  ) {}

  // ---------------------------------------------------------------- market

  async getQuotes(symbols: string[]): Promise<Quote[]> {
    const body = await this.request("GET", `/v1/market/quotes?symbols=${symbols.join(",")}`);
    const quotes: Quote[] = [];
    for (const q of body.quotes as Record<string, unknown>[]) {
      // NOT_FOUND / NO_DATA 는 개별 종목의 정상 결과 — 가격이 없으므로 제외한다 (ctx.quote() 가 undefined)
      if (q.outcome !== "OK" || q.price == null) continue;
      quotes.push({
        symbol: String(q.symbol),
        price: D(q.price),
        bidPrice: DorNull(q.bidPrice),
        askPrice: DorNull(q.askPrice),
        volume: Number(q.volume ?? 0),
        change: DorNull(q.change),
        changeRate: pct(q.changeRate),
        timestamp: parseKst(q.lastTradeAt) ?? parseKst(q.requestedAt) ?? new Date(),
      });
    }
    return quotes;
  }

  async getCandles(symbol: string, interval: CandleInterval, limit?: number): Promise<Candle[]> {
    let path = `/v1/market/candles?symbol=${symbol}&interval=${interval}`;
    if (limit !== undefined) path += `&limit=${limit}`;
    const body = await this.request("GET", path);
    return (body.candles as Record<string, unknown>[]).map((c) => ({
      timestamp: parseKst(c.time) ?? new Date(0),
      open: D(c.open), high: D(c.high), low: D(c.low), close: D(c.close),
      volume: Number(c.volume ?? 0),
    }));
  }

  /** v1.3: date 는 거래소 현지 일자, 세션 시각은 KST → 뉴욕 현지 HH:mm 으로 바꿔 담는다 */
  async getCalendar(): Promise<MarketDay[]> {
    const body = await this.request("GET", "/v1/market/calendar");
    return (body.calendar as Record<string, unknown>[]).map((d) => {
      const sessions = (d.sessions as Record<string, unknown>[] | null) ?? [];
      const regular = sessions.find((s) => s.type === "REGULAR");
      const start = regular ? nyClock(regular.open) : null;
      const end = regular ? nyClock(regular.close) : null;
      const open = (d.status === "OPEN" || d.status === "HALF_DAY") && start !== null && end !== null;
      return {
        date: String(d.date),
        open,
        regular: open ? { start: start!, end: end! } : null,
        timezone: NEW_YORK,
        holiday: (d.holidayName as string | null) ?? null,
      };
    });
  }

  // --------------------------------------------------------------- account

  /** v1.3 계좌 응답은 예수금(cashAmount)만 준다. 총평가는 예수금 + 보유 평가금액 합 (보유 조회 1회 추가) */
  async getAccount(): Promise<Account> {
    const body = await this.request("GET", "/v1/account", { account: true });
    const cash = D(body.cashAmount);
    const holdings = await this.getHoldings();
    const marketValue = holdings.reduce((acc, h) => acc.plus(h.marketValue ?? 0), new Decimal(0));
    return {
      accountId: String(body.accountId),
      currency: String(body.currency ?? "USD"),
      cash,
      portfolioValue: cash.plus(marketValue),
      status: "ACTIVE",
      name: null,
    };
  }

  async getHoldings(): Promise<Holding[]> {
    const body = await this.request("GET", "/v1/account/holdings", { account: true });
    return (body.holdings as Record<string, unknown>[]).map((h) => ({
      symbol: String(h.symbol),
      quantity: D(h.quantity),
      avgEntryPrice: D(h.averageBuyPrice),
      currentPrice: DorNull(h.currentPrice),
      marketValue: DorNull(h.evaluationAmount),
      unrealizedPnl: DorNull(h.evaluationPnl),
      unrealizedPnlRate: pct(h.evaluationPnlRate),
    }));
  }

  async getBuyingPower(): Promise<Decimal> {
    const body = await this.request("GET", "/v1/account/buying-power", { account: true });
    return D(body.buyingPower);
  }

  // ---------------------------------------------------------------- orders

  async createOrder(request: CreateOrderRequest): Promise<Order> {
    const payload: Record<string, unknown> = {
      // v1.3: clientOrderId(멱등키)·market 필수 — 호출자가 안 주면 어댑터가 UUID 를 만든다
      clientOrderId: request.clientOrderId || randomUUID(),
      market: MARKET,
      symbol: request.symbol,
      side: request.side,
      orderType: request.orderType,
      quantity: request.quantity.toString(),
      timeInForce: request.timeInForce ?? "DAY",
    };
    if (request.limitPrice) payload.limitPrice = request.limitPrice.toString();
    return this.toOrder(await this.request("POST", "/v1/orders", { account: true, json: payload }));
  }

  async getOrders(): Promise<Order[]> {
    const body = await this.request("GET", "/v1/orders", { account: true });
    return (body.orders as Record<string, unknown>[]).map((o) => this.toOrder(o));
  }

  async getOrder(orderId: string): Promise<Order> {
    return this.toOrder(await this.request("GET", `/v1/orders/${orderId}`, { account: true }));
  }

  async cancelOrder(orderId: string): Promise<Order> {
    return this.toOrder(await this.request("DELETE", `/v1/orders/${orderId}`, { account: true }));
  }

  async getFills(): Promise<Fill[]> {
    const body = await this.request("GET", "/v1/orders/fills", { account: true });
    return (body.fills as Record<string, unknown>[]).map((f) => ({
      fillId: null, // v1.3: 원장이 체결 ID 를 발급하지 않는다
      orderId: (f.orderId as string) ?? null,
      symbol: (f.symbol as string) ?? null,
      side: (f.side as OrderSide) ?? null,
      quantity: DorNull(f.quantity),
      price: DorNull(f.price),
    }));
  }

  // -------------------------------------------------------------- internal

  /** 생성/상세/취소 응답 공통 — 생성·취소는 orderId/status/requestedAt 만 온다 */
  private toOrder(body: Record<string, unknown>): Order {
    const status = ORDER_STATUSES.includes(String(body.status)) ? (body.status as OrderStatus) : "UNKNOWN";
    const orderType = String(body.orderType ?? "").toUpperCase();
    return {
      orderId: String(body.orderId),
      status,
      symbol: (body.symbol as string) ?? null,
      side: (body.side as OrderSide) ?? null,
      orderType: orderType === "MARKET" || orderType === "LIMIT" ? (orderType as OrderType) : null,
      quantity: DorNull(body.quantity),
      limitPrice: DorNull(body.limitPrice),
      filledQuantity: DorNull(body.filledQuantity),
      avgFillPrice: DorNull(body.avgFillPrice),
      clientOrderId: (body.requestId as string) ?? null, // v1.3 상세 조회의 clientOrderId 필드명
      submittedAt: parseKst(body.requestedAt),
      canceledAt: null,
    };
  }

  private async request(
    method: string, path: string,
    opts: { account?: boolean; json?: Record<string, unknown> } = {},
  ): Promise<Record<string, unknown>> {
    const call = async (): Promise<Record<string, unknown>> => {
      const headers: Record<string, string> = {
        Authorization: `Bearer ${await this.getToken()}`,
        [NEXT_REQUEST_ID_HEADER]: newRequestId(),
      };
      if (opts.account) headers[NEXT_ACCOUNT_HEADER] = this.accountId;
      if (opts.json) headers["Content-Type"] = "application/json";
      const [status, body] = await httpJson(this.baseUrl + path, {
        method, headers, body: opts.json ? JSON.stringify(opts.json) : undefined,
      });
      if (status < 200 || status >= 300) throw this.mapError(status, (body.error ?? {}) as NextError);
      return body;
    };
    try {
      return await call();
    } catch (e) {
      if (e instanceof AuthError) {
        this.token = null; // 토큰 만료 - 1회 재발급 후 재시도
        return call();
      }
      throw e;
    }
  }

  /**
   * v1.3 에러 type ↔ HTTP: validation(400) authentication(401) permission(403) not_found(404)
   * conflict(409) business_rule(422) locked(423, 킬스위치 trading-halted) rate_limit(429) server(5xx)
   */
  private mapError(status: number, error: NextError): BrokerApiError {
    const code = error.code ?? null;
    const message = `Next(${code}) ${error.message ?? ""} requestId=${error.requestId}`;
    if (status === 401 || error.type === "authentication") return new AuthError(status, code, message);
    if (status === 429) return new RateLimitError(status, code, message);
    // 조회전용 키(insufficient-scope)·계좌 불일치 — 자금 부족으로 오인하지 않는다
    if (error.type === "permission") return new BrokerApiError(status, code, message);
    if (code === "order-not-found") return new OrderNotFoundError(code, message);
    if (code?.includes("insufficient")) return new InsufficientFundsError(status, code, message);
    if (code === "trading-halted" || code?.includes("market-closed")) return new MarketClosedError(status, code, message);
    if (error.type === "validation" || error.type === "business_rule") return new InvalidOrderError(status, code, message);
    return new BrokerApiError(status, code, message);
  }

  private async getToken(): Promise<string> {
    if (this.token && Date.now() < this.tokenExpiresAt - 60_000) return this.token;
    const form = new URLSearchParams({
      grant_type: "client_credentials",
      client_id: this.clientId,
      client_secret: this.clientSecret,
    });
    const [status, body] = await httpJson(`${this.baseUrl}/v1/oauth/token`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        [NEXT_REQUEST_ID_HEADER]: newRequestId(),
      },
      body: form.toString(),
    });
    if (status !== 200 || typeof body.access_token !== "string") throw this.tokenError(status, body);
    this.token = body.access_token;
    this.tokenExpiresAt = Date.now() + Number(body.expires_in ?? 43200) * 1000;
    return this.token;
  }

  private tokenError(status: number, body: Record<string, unknown>): AuthError {
    // 토큰 발급 400/401 은 OAuth 표준 형식: {"error":"invalid_client","error_description":"..."}
    if (typeof body.error === "string") {
      return new AuthError(status, body.error, `Next 토큰 발급 실패(${body.error}): ${body.error_description ?? ""}`);
    }
    // 429/5xx 는 플랫폼 엔벨로프: {"error":{"code":..,"message":..}}
    const error = (body.error ?? {}) as NextError;
    return new AuthError(status, error.code ?? null, `Next 토큰 발급 실패(${error.code}): ${error.message ?? ""}`);
  }
}
